Plutarch/Internal/Lift.hs

```diff
@@ -0,0 +1,263 @@
+differently somehow, use 'fromPlutarch'.
+
+@since WIP
+-}
+plift ::
+  forall (a :: S -> Type).
+  PLiftable a =>
+  (forall (s :: S). Term s a) ->
+  AsHaskell a
+plift t = case fromPlutarch (PLifted @a (punsafeCoerce t)) of
+  Left err ->
+    error $
+      "plift failed: "
+        <> ( case err of
+              CouldNotEvaluate evalErr -> "term errored: " <> show evalErr
+              TypeError builtinError -> "incorrect type: " <> show builtinError
+              CouldNotCompile compErr -> "could not compile: " <> Text.unpack compErr
+              CouldNotDecodeData -> "Data value is not a valid encoding for this type"
+           )
+  Right res -> res
+
+{- | @via@-deriving helper, indicating that @a@ has a Haskell-level equivalent
+@h@ that is directly part of the Plutus default universe (instead of by way
+of a @Data@ encoding).
+
+@since WIP
+-}
+newtype DeriveBuiltinPLiftable (a :: S -> Type) (h :: Type) (s :: S)
+  = DeriveBuiltinPLiftable (a s)
+  deriving stock (Generic)
+  deriving anyclass (PlutusType)
+
+-- | @since WIP
+instance DerivePlutusType (DeriveBuiltinPLiftable a h) where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since WIP
+instance
+  ( PlutusType a
+  , PLC.DefaultUni `Includes` h
+  ) =>
+  PLiftable (DeriveBuiltinPLiftable a h)
+  where
+  type AsHaskell (DeriveBuiltinPLiftable a h) = h
+  {-# INLINEABLE toPlutarch #-}
+  toPlutarch = unsafeUniToPLifted
+  {-# INLINEABLE fromPlutarch #-}
+  fromPlutarch = unsafePLiftedToUni
+
+{- | @via@-deriving helper, indicating that @a@ has a Haskell-level equivalent
+@h@ by way of its @Data@ encoding, rather than by @h@ being directly part of
+the Plutus default universe.
+
+@since WIP
+-}
+newtype DeriveDataPLiftable (a :: S -> Type) (h :: Type) (s :: S)
+  = DeriveDataPLiftable (a s)
+  deriving stock (Generic)
+  deriving anyclass (PlutusType)
+
+-- | @since WIP
+instance DerivePlutusType (DeriveDataPLiftable a h) where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since WIP
+instance
+  ( PlutusType a
+  , PInner a ~ PData
+  , PTx.ToData h
+  , PTx.FromData h
+  ) =>
+  PLiftable (DeriveDataPLiftable a h)
+  where
+  type AsHaskell (DeriveDataPLiftable a h) = h
+  {-# INLINEABLE toPlutarch #-}
+  toPlutarch = unsafeUniToPLifted @_ @PLCData.Data . PTx.toData
+  {-# INLINEABLE fromPlutarch #-}
+  fromPlutarch t = do
+    res <- unsafePLiftedToUni t
+    case PTx.fromData res of
+      Nothing -> Left CouldNotDecodeData
+      Just res' -> pure res'
+
+{- | Helper for writing 'PLifted' instances. This is /highly/ unsafe, and only
+suitable for internal use!
+
+@since WIP
+-}
+unsafeUniToPLifted ::
+  forall (a :: S -> Type) (h :: Type) (s :: S).
+  PLC.DefaultUni `Includes` h =>
+  h ->
+  PLifted a s
+unsafeUniToPLifted =
+  PLifted . punsafeConstantInternal . PLC.someValue @h @PLC.DefaultUni
+
+{- | Helper for writing 'PLiftable' instances. This is /highly/ unsafe, and only
+suitable for internal use!
+
+@since WIP
+-}
+unsafePLiftedToUni ::
```

Comment:
Nitpick. We have `[to|from]Plutarch` but this is "to" both ways. It'd be nice to keep naming consistent

Reply:
So `unsafePLiftedToUni` and `unsafePLiftedFromUni`?